nixos/nginx: add strictTransportSecurity options

bmillwood opened the pull request:
Let me know if this needs a changelog entry, I wasn't sure whether it was significant enough.

This is my first nontrivial nixpkgs PR, so sorry about any mess :) I didn't `nixfmt` these modules because it would change a lot more than what I touched.

## Things done

- Built on platform(s)
  - [X] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [X] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) (or backporting [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) and [24.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [X] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

bmillwood committed:
nixos/nginx: add strictTransportSecurity options

This is pretty easy for the user to set themselves, but it's a
recommended security setting so making it more discoverable / encouraged
seems like a good thing.

emilazy commented on nixos/modules/services/web-servers/nginx/vhost-options.nix:
```diff
@@ -212,6 +212,86 @@ with lib;
+
+      includeSubdomains = mkOption {
+        type = types.bool;
+        default = true;
```
I am not sure this should be on by default, it is a footgun unless you are really prepared. Lots of cases of random sub‐subdomains with internal test deployments or services that aren’t ready for it.

(Haven’t done a thorough review of the rest of the PR yet.)

bmillwood replied:
sure, done :)

tbh my attitude was "enableACME is so easy these days" but I guess non-public uses have the potential to be trickier

emilazy replied:
Yeah, the problem is that it‘s easy if you have one single NixOS machine serving your entire domain hierarchy, and it‘s easy if you have a trillion machines that are all behind consistent HTTPS frontend stuff proxying everything to your highly‐structured web scale enterprise backends, but there‘s a wide range of setups between those where you just don‘t have complete visibility and control into everything going on under `*.mycorp.com`. Intranet, legacy SaaS, a random appliance you can‘t update, … there’s a big gap between “this web server on this domain can reliably do HTTPS” and “absolutely nothing under this hierarchy exists that cannot reliably do HTTPS in every single context”, and with HSTS it’s easy to screw up painfully. Of course people should set it if they can.

Will try to take a more detailed look at this PR in the next few days if nobody beats me to it, feel free to ping if I forget.

bmillwood committed:
nixos/nginx: HSTS includeSubdomains default false

For more-or-less the same reasons that HSTS defaults false: accidentally
enabling it when you didn't intend to can be really annoying.

bmillwood commented:
Something else I considered: should we have any warnings that e.g. complain if `strictTransportSecurity.enable && !hasSSL` ? I leant in the direction of trusting the user and imagining there might be some weird combination of reverse proxies and so on where it makes sense to send the header even though SSL termination is somewhere else in the path, but I also want to be helpful as far as possible.

bmillwood commented:
whoever reviews this: is it OK to just tag a random apache-httpd reviewer (or reviewers) to ask if they want the same to be added there? (which also prompts the question: should we be sharing code between the two / plus maybe caddy / others as well? my guess is that code sharing usually isn't helpful and it doesn't make sense to bother doing it for those few cases where it would work, but idk the idea of having all webserver configs look ~ the same is cool)

SuperSandro2000 commented on nixos/modules/services/web-servers/nginx/vhost-options.nix:
```diff
@@ -212,6 +212,86 @@ with lib;
+        type = types.int;
+        default = 63072000;
+        description = ''
+          If HTTP Strict Transport Security (HSTS) is enabled by
+          `strictTransportSecurity.enable`, this is how long the browser is
+          instructed to retain the setting, in seconds. 2 years (63072000 seconds)
+          is recommended, but if you're introducing HSTS for the first time on an
+          existing site you may want to ramp up the value gradually, or set it to
```
```suggestion
          This is how long the browser is instructed to retain the HSTS setting in seconds.
          2 years (63072000 seconds) is recommended, but if you're introducing HSTS for the first time on an
          existing site you may want to ramp up the value gradually, or set it to
```

SuperSandro2000 commented on nixos/modules/services/web-servers/nginx/vhost-options.nix:
```diff
@@ -212,6 +212,86 @@ with lib;
+        type = types.bool;
+        default = false;
+        description = ''
+          If HTTP Strict Transport Security (HSTS) is enabled by
+          `strictTransportSecurity.enable`, also specify the `includeSubdomains`
+          directive, so that e.g. if the vhost is https://www.website.example it
+          also sets the HSTS policy for https://sub.www.website.example. (Note
+          that "sibling" domains like https://mail.website.example are not
+          affected.)
```
```suggestion
          Also specify the `includeSubdomains` directive, so that e.g. if the virtual host is `https://www.example.com` it
          also sets the HSTS policy for `https://sub.www.example.com`
          
          ::: {.note}
          "Sibling" domains like https://mail.example.com are not affected.
          :::
```

SuperSandro2000 commented on nixos/modules/services/web-servers/nginx/vhost-options.nix:
```diff
@@ -212,6 +212,86 @@ with lib;
+        # default
+        default = false;
+        description = ''
+          If HTTP Strict Transport Security (HSTS) is enabled by
+          `strictTransportSecurity.enable`, also specify the (non-standard, but
+          widely recognised) `preload` directive. This allows submitting your
+          site to https://hstspreload.org/ to be added to the HSTS preload list
+          used by all major browsers, so that it is not necessary to visit your
+          site even once to enable HSTS.
```
```suggestion
          Also specify the `preload` directive. This allows submitting your
          site to https://hstspreload.org/ to be added to the HSTS preload list
          used by all major browsers, so that it is not necessary to visit your
          site even once to enable HSTS.
```

SuperSandro2000 commented on nixos/modules/services/web-servers/nginx/vhost-options.nix:
```diff
@@ -212,6 +212,86 @@ with lib;
+        type = types.bool;
+        default = false;
+        description = ''
+          Whether to enable HTTP Strict Transport Security (HSTS). This sends
+          the header "strict-transport-security: max-age=63072000" on all
+          responses, which tells clients to refuse to use an insecure HTTP
```
This depends on the other options and we shouldn't repeat the defauls of the other options here

SuperSandro2000 commented on nixos/modules/services/web-servers/nginx/vhost-options.nix:
```diff
@@ -212,6 +212,86 @@ with lib;
+      };
+
+      seconds = mkOption {
+        type = types.int;
```
```suggestion
        type = types.ints.positive;
```

bmillwood replied:
I think 0 is also an allowed value here, and makes some sense to ask clients to clear HSTS settings, so `types.ints.unsigned`?

bmillwood replied:
(I switched to `unsigned`, but lmk if you disagree)

bmillwood committed:
nixos/nginx: more concise HSTS option descriptions

Co-authored-by: Sandro <sandro.jaeckel@gmail.com>

bmillwood committed:
nixos/nginx: more HSTS option changes from review

- seconds can't be negative
- tweak wording + fix whitespace errors in docs

bmillwood committed:
nixos/nginx: don't duplicate HSTS option defaults

bmillwood committed:
nixos/nginx: cautionary note about HSTS includeSubdomains

bmillwood commented on nixos/modules/services/web-servers/nginx/vhost-options.nix:
```diff
@@ -212,6 +212,86 @@ with lib;
+          responses, which tells clients to refuse to use an insecure HTTP
+          connection for this host for the next 63072000 seconds (approximately
+          2 years), or for the duration specified by
+          `strictTransportSecurity.seconds` if set. This helps protects against
```
(I believe this is covered by the change I made for the other comment)

bmillwood commented:
@SuperSandro2000 are you happy with this PR after my changes?

bmillwood committed:
treefmt

Ma27 reviewed (changes requested):
I think this option is a bad idea: if you enable this and then try to `add_header` on a location-level, this will silently be discarded which is a bad interface IMHO.
